editor/user_stats.hpp

```diff
@@ -1,30 +1,66 @@
 #pragma once
 
 #include "std/cstdint.hpp"
+#include "std/ctime.hpp"
+#include "std/function.hpp"
+#include "std/mutex.hpp"
+#include "std/shared_ptr.hpp"
 #include "std/string.hpp"
 
 namespace editor
 {
 class UserStats
 {
 public:
-  explicit UserStats(string const & userName);
+  UserStats(time_t const updateTime, uint32_t const rating, uint32_t const changesCount)
+    : m_changesCount(changesCount), m_rank(rating), m_updateTime(updateTime)
+  {
+  }
 
-  bool IsChangesCountInitialized() const;
-  bool IsRankInitialized() const;
+  bool GetChangesCount(int32_t & changesCount) const;
+  bool GetRank(int32_t & rank) const;
 
-  int32_t GetChangesCount() const { return m_changesCount; }
-  int32_t GetRank() const { return m_rank; }
-
-  bool GetUpdateStatus() const { return m_updateStatus; }
-  bool Update();
+  time_t GetLastUpdate() const { return m_updateTime; }
 
 private:
-  string m_userName;
   int32_t m_changesCount;
   int32_t m_rank;
+  time_t m_updateTime;
+};
+
+class UserStatsLoader
+{
+public:
+  UserStatsLoader();
+
+  using TOnUpdateCallback = function<void()>;
+
+  /// Synchronously sends request to the server. Updates stats and returns true on success.
+  bool Update(string const & userName);
+  /// Launch the update process if stats are too old.
+  /// The process posts fn to a gui thread on success.
+  void Update(string const & userName, TOnUpdateCallback const & fn);
+
+  /// Resets internal state and removes records from settings.
+  void DropStats(string const & userNamae);
+
+  /// Atomically returns stats if userName is still actual.
+  shared_ptr<UserStats const> GetStats(string const & userName) const;
+
+  /// Debug only.
+  string GetUserName() const;
+
+private:
+  bool LoadFromSettings();
+  /// Not thread-safe, use synchonization.
+  void SaveToSettings();
+  void DropSettings();
+
+  string m_userName;
+
+  time_t m_lastUpdate{};
+  mutable mutex m_mutex;
 
-  /// True if last update was successful.
-  bool m_updateStatus;
+  shared_ptr<UserStats> m_userStats;
```

Comment:
Why do you need shared_ptr here? Just keep it by value, and return it by value where it's needed.